### PR TITLE
fix(index): delete skipped documents for StructuredDoc

### PR DESCRIPTION
**Pull request by zwpaper**
created a pr:
CleanShot X 2024-11-25 13 50 13

close PR and reindex logs:
```
2024-11-25T14:22:13.800883+08:00 [INFO]: Pulling source code for repository zwpaper/testing
2024-11-25T14:22:13.800956+08:00 [INFO]: Building source code index: https://github.com/zwpaper/testing
2024-11-25T14:22:15.303840+08:00 [INFO]: Processed 2/2 files, updated 0 chunks
2024-11-25T14:22:15.307414+08:00 [INFO]: Finished garbage collection for code index: 6 items kept, 0 items removed
2024-11-25T14:22:15.319007+08:00 [INFO]: Indexing documents for repository zwpaper/testing
2024-11-25T14:22:17.009773+08:00 [INFO]: 1 docs seen, 0 docs updated, 1 docs deleted
2024-11-25T14:22:17.027753+08:00 [INFO]: Job completed successfully
```

after PR closed:
CleanShot X 2024-11-25 14 22 37

### CODE REVIEW

File: ee/tabby-webserver/src/service/background_job/third_party_integration.rs [hunk 1]
```diff
@@ -140,18 +140,23 @@ impl SchedulerGithubGitlabJob {
         stream! {
             let mut count = 0;
             let mut num_updated = 0;
-            for await (updated_at, doc) in issue_stream.chain(pull_stream) {
-                if index.add(updated_at, doc).await {
+            let mut num_deleted = 0;
+            for await (state, doc) in issue_stream.chain(pull_stream) {
+                let id = &doc.id().to_owned();
+                if !state.should_clean && index.add(state.updated_at, doc).await {
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/background_job/third_party_integration.rs:146]:
```suggestion
                if !index.add(state, doc).await {
```

let indexer to decide whether to delete document based on `state`

File: ee/tabby-webserver/src/service/background_job/third_party_integration.rs [hunk 2]
```diff
@@ -140,18 +140,23 @@ impl SchedulerGithubGitlabJob {
...
                     num_updated += 1
                 }
+                if state.should_clean && index.delete(id).await {
+                    num_deleted += 1;
+                }
 
                 count += 1;
                 if count % 100 == 0 {
-                    logkit::info!("{} docs seen, {} docs updated", count, num_updated);
+                    logkit::info!("{} docs seen, {} docs updated, {} docs deleted", count, num_updated, num_deleted);
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/background_job/third_party_integration.rs:155]: revert

File: ee/tabby-webserver/src/service/background_job/third_party_integration.rs [hunk 3]
```diff
@@ -140,18 +140,23 @@ impl SchedulerGithubGitlabJob {
...
                 };
             }
 
-            logkit::info!("{} docs seen, {} docs updated", count, num_updated);
+            logkit::info!("{} docs seen, {} docs updated, {} docs deleted", count, num_updated, num_deleted);
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/background_job/third_party_integration.rs:159]: ditto

File: ee/tabby-webserver/src/service/background_job/third_party_integration.rs [hunk 4]
```diff
@@ -179,11 +184,16 @@ impl SchedulerGithubGitlabJob {
     }
 }
 
+pub struct FetchState {
+    updated_at: DateTime<Utc>,
+    should_clean: bool,
+}
```
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/background_job/third_party_integration.rs:190]:
```suggestion
pub StructuredDocState  {
    updated_at: DateTime<Utc>,
    deleted: bool,
}
```

Move definition to `tabby-index/src/structured_doc/public.rs`

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: crates/tabby-index/src/structured_doc/public.rs [hunk 5]
```diff
@@ -14,6 +14,11 @@ pub use super::types::{
 use super::{create_structured_doc_builder, types::BuildStructuredDoc};
 use crate::{indexer::TantivyDocBuilder, Indexer};
 
+pub struct StructuredDocState {
+    pub updated_at: DateTime<Utc>,
+    pub deleted: bool,
```
Comment (wsxiaoys) [inline crates/tabby-index/src/structured_doc/public.rs:19]: add field level comments to explain the behavior.

File: crates/tabby-index/src/structured_doc/public.rs [hunk 6]
```diff
@@ -14,8 +14,14 @@ pub use super::types::{
 use super::{create_structured_doc_builder, types::BuildStructuredDoc};
 use crate::{indexer::TantivyDocBuilder, Indexer};
 
+/// StructuredDocState is used to track the state of the document source.
+/// It is used to determine whether the document should be updated or deleted.
 pub struct StructuredDocState {
+    // updated_at is the time when the document was last updated.
```
Comment (wsxiaoys) [inline crates/tabby-index/src/structured_doc/public.rs:20]: please explain how it's used to incremental update the document in index.

File: crates/tabby-index/src/structured_doc/public.rs [hunk 7]
```diff
@@ -14,8 +14,14 @@ pub use super::types::{
...
     pub updated_at: DateTime<Utc>,
+    // deleted indecates whether the document should be deleted in indexer
```
Comment (wsxiaoys) [inline crates/tabby-index/src/structured_doc/public.rs:22]: fix typo

Comment (zwpaper) [reply crates/tabby-index/src/structured_doc/public.rs:22]: which word?

Comment (wsxiaoys) [reply crates/tabby-index/src/structured_doc/public.rs]: `indecates`

in future i recommend always improve text copy withllm models.

**Pull request merged**